### DISCUSSION TITLE
PCF8574 Mode 2 per script

### DISCUSSION THREAD

**Discussion by DC-555** [arendst/Tasmota#18934] [Q&A]
I would like to put my PCF8574 module in mode 2 via script, but I couldn't do it anyway.
I already have a script and would like to include my model along with it:

Model: {"NAME":"PCF8574","BASE":0,"GPIO":[192,193,194,195,258,259,0,0]}

The first 4 pins will act as an input switch, the other 2 as a relay output, and the last ones disabled.

I currently have this script template:

>B
; ### EXECUTE ON BOOT ###

ini=1
; Signals initialization

if ((ini==1) and ((pwr[1]==1) or (pwr[2]==1)))
then
=>POWER0 0
endif
; Routine reset outputs at boot

user_config_override.h :

//-- PCF8574 configuration -----------------------
#ifndef USE_PCF8574
#define USE_PCF8574
#endif
#define USE_PCF8574_MODE2
#define USE_PCF8574_SENSOR
#define USE_PCF8574_DISPLAYINPUT
#define USE_PCF8574_MQTTINPUT
#define PCF8574_ADDR1 0x20
//------------------------------------------------ ---

I've tried to put it like this in the manual but without success:

if not found and scripts are supported it looks for a specific script like -y <template>

**Comment by arendst** [comment]
@gemu2015 I'm trying to use ``>y`` in a script to load PCF8574 JSON config data. I'm not sure how to use this in a script as I never did scripts.

What the PCF8574 driver does is calling ``ScriptLoadSection(">y")`` hoping to get a return string holding ``{"NAME":"PCF8574","BASE":0,"GPIO":[192,193,194,195,256,257,0,0]}``.

If I use a script like this:
```
>D
>y{"NAME":"PCF8574","BASE":0,"GPIO":[192,193,194,195,256,257,0,0]}#
```

It doesn't work. Any help appreciated.

**Reply by gemu2015** [reply]
\>y must be on a separate line, try this
```
>y
{"NAME":"PCF8574","BASE":0,"GPIO":[192,193,194,195,256,257,0,0]}
#
```

**Reply by arendst** [reply]
Nope, no change:
```
00:00:00.001 HDW: ESP8266EX
00:00:00.050-032 CFG: Loaded from flash at FA, Count 663
00:00:00.056-032 QPC: Count 1
00:00:00.061-032 I2C: PCF8574 found at 0x20
00:00:00.064-032 PCF: Total devices 1, PCF8574 output ports 0
00:00:00.065-032 MCP: Invalid template
00:00:00.072-032 ROT: Mode 1
00:00:00.074-032 SRC: Restart
00:00:00.075-032 Project tasmota - Wemos5 Version 13.0.0.1(theo)-2_7_4_9(2023-06-29T16:55:44)
00:00:00.079-029 Script: nv=0, tv=0, vns=0, vmem=16, smem=2560
00:00:00.187-029 WIF: Attempting connection...
00:00:00.523-029 WIF: Connecting to AP2 indebuurt_IoT Channel 11 BSSId 18:E8:29:CA:17:C1 in mode 11n as wemos5...
00:00:01.755-029 WIF: Connected
00:00:02.009-026 HTP: Web server active on wemos5 with IP address 192.168.2.143
00:00:02.498-026 RTC: UTC 2023-06-29T15:41:22, DST 2023-03-26T02:00:00, STD 2023-10-29T03:00:00
17:41:22.000-026 RTC: Synced by NTP
17:41:23.008-026 MQT: Attempting connection...
17:41:23.022-026 MQT: Connected
17:41:23.026-023 MQT: tele/wemos5/LWT = Online (retained)
17:41:23.029-024 MQT: cmnd/wemos5/POWER = 
17:41:23.031-024 MQT: Subscribe to cmnd/wemos5/#
17:41:23.034-024 MQT: Subscribe to cmnd/tasmotas/#
17:41:23.038-024 MQT: Subscribe to cmnd/DVES_17E067_fb/#
17:41:23.043-024 MQT: tele/wemos5/INFO1 = {"Info1":{"Module":"Generic","Version":"13.0.0.1(theo)","FallbackTopic":"cmnd/DVES_17E067_fb/","GroupTopic":"cmnd/tasmotas/"}}
17:41:23.057-024 MQT: tele/wemos5/INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"wemos5","IPAddress":"192.168.2.143"}}
17:41:23.067-024 MQT: tele/wemos5/INFO3 = {"Info3":{"RestartReason":"Software/System restart","BootCount":252}}
17:41:26.486-026 QPC: Reset
17:41:27.450-022 MQT: tele/wemos5/STATE = {"Time":"2023-06-29T17:41:27","Uptime":"0T00:00:09","UptimeSec":9,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":2,"SSId":"indebuurt_IoT","BSSId":"18:E8:29:CA:17:C1","Channel":11,"Mode":"11n","RSSI":100,"Signal":-36,"LinkCount":1,"Downtime":"0T00:00:03"}}
17:41:28.448-025 APP: Boot Count 252
17:41:28.620-025 CFG: Saved to flash at F9, Count 664, Bytes 4096
```
where it should hev looked like this (using rules):
```
00:00:00.001 HDW: ESP8266EX
00:00:00.053-032 CFG: Loaded from flash at F9, Count 623
00:00:00.059-032 QPC: Count 1
00:00:00.065-032 I2C: PCF8574 found at 0x20
00:00:00.068-032 PCF: Base 0, Template 'PCF8574'
00:00:00.071-032 MCP: Invalid template
00:00:00.076-032 SWT: Add vSwitch1, State 1
00:00:00.077-032 SWT: Add vSwitch2, State 1
00:00:00.078-032 SWT: Add vSwitch3, State 1
00:00:00.079-032 SWT: Add vSwitch4, State 1
00:00:00.084-032 ROT: Mode 1
00:00:00.086-032 SRC: Restart
00:00:00.088-032 Project tasmota - Wemos5 Version 13.0.0.1(theo)-2_7_4_9(2023-06-29T16:44:15)
00:00:00.199-032 WIF: Attempting connection...
00:00:00.535-032 WIF: Connecting to AP2 indebuurt_IoT Channel 11 BSSId 18:E8:29:CA:17:C1 in mode 11n as wemos5...
00:00:01.755-032 WIF: Connected
00:00:02.007-029 HTP: Web server active on wemos5 with IP address 192.168.2.143
00:00:02.462-028 RTC: UTC 2023-06-29T14:49:49, DST 2023-03-26T02:00:00, STD 2023-10-29T03:00:00
16:49:49.000-028 RTC: Synced by NTP
16:49:49.028-028 MQT: Attempting connection...
16:49:49.042-028 MQT: Connected
16:49:49.045-025 MQT: tele/wemos5/LWT = Online (retained)
16:49:49.048-027 MQT: cmnd/wemos5/POWER = 
16:49:49.051-027 MQT: Subscribe to cmnd/wemos5/#
16:49:49.054-027 MQT: Subscribe to cmnd/tasmotas/#
16:49:49.058-027 MQT: Subscribe to cmnd/DVES_17E067_fb/#
16:49:49.064-027 MQT: tele/wemos5/INFO1 = {"Info1":{"Module":"Generic","Version":"13.0.0.1(theo)","FallbackTopic":"cmnd/DVES_17E067_fb/","GroupTopic":"cmnd/tasmotas/"}}
16:49:49.077-026 MQT: tele/wemos5/INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"wemos5","IPAddress":"192.168.2.143"}}
16:49:49.087-026 MQT: tele/wemos5/INFO3 = {"Info3":{"RestartReason":"Software/System restart","BootCount":236}}
16:49:49.096-025 MQT: stat/wemos5/POWER1 = {"POWER1":"OFF"}
16:49:49.100-025 MQT: stat/wemos5/POWER1 = OFF
16:49:49.104-025 MQT: stat/wemos5/POWER2 = {"POWER2":"OFF"}
16:49:49.109-025 MQT: stat/wemos5/POWER2 = OFF
16:49:52.489-028 QPC: Reset
```

**Reply by arendst** [reply]
When I save the script I find it strange ``smem`` isn't changing:

Script contains >D only:
```
17:43:33.682-024 HTP: Edit Script
17:43:57.152-023 script compressed to 4 bytes = 133 %
17:43:57.154-023 Script: nv=0, tv=0, vns=0, vmem=16, smem=2560
17:43:57.156-023 HTP: Consoles
17:43:57.570-023 CFG: Saved to flash at F8, Count 665, Bytes 4096
```
Now adding 
```
>y
{"NAME":"PCF8574","BASE":0,"GPIO":[192,193,194,195,256,257,0,0]}
#
```
and it shows:
```
17:44:21.284-023 HTP: Edit Script
17:44:28.347-023 script compressed to 75 bytes = 104 %
17:44:28.349-022 Script: nv=0, tv=0, vns=0, vmem=16, smem=2560
17:44:28.351-023 HTP: Consoles
17:44:28.608-023 CFG: Saved to flash at F7, Count 666, Bytes 4096
```

**Reply by arendst** [reply]
I fear the issue is related to the init sequence. The PCF8574 tries to retreive the data BEFORE scripts are initialized. This doesn't work as the ``scriptptr`` isn't initialzed.

I'll try to add a read from settings memory without first initializing scripts.

**Comment by DC-555** [comment]
As a solution, I developed a script that downloads the pcf8574.dat file once.

To make the link I created a file on github with the same name, then I copied the link from the "raw" file.
The link can also be sent in CMD or MQTT Command
The download link is this:
br def urlfetch(url,file); if file==nil; import string; file=string.split(url,'/').pop(); end; var wc=webclient(); wc.begin(url); var st=wc.GET(); if st!=200 raise 'connection_error','status: '+str(st) end; st='Fetched '+str(wc.write_file(file)); print(url,st); wc.close(); return st; end; urlfetch('https://raw.githubusercontent.com/DC-555/pcf8574.dat/287a9f04fb20647ad2c70693c8b5d1c2d323f195/pcf8574.dat')

**Reply by DC-555** [reply]
You can also try:
#define START_SCRIPT_FROM_BOOT

**Comment by arendst** [comment]
@gemu2015 I can get it to work by changing scripter init sequence.

You currently call the init sequence at ``FUNC_INIT`` at line 12513. If I change this to ``FUNC_SETUP_RING1`` AND re-enable the commented ``FUNC_INIT`` at lines 12675/6 then it works.

Regarding filesystem and eeprom support I see no issue with this change as both filesystem AND I2C are (just) initialized before ``FUNC_SETUP_RING1``.

Can you verify this is a solid solution for scripter?

**Reply by gemu2015** [reply]
ok, i am very busy today with other things. will verify at the weekend

**Reply by arendst** [reply]
Thx. For now I merged the change allowing easier test for you.

**Reply by gemu2015** [reply]
ok, since sd card is not ready at this state script on sd  card are no longer recognized.
many of my devices use sd cards with scripts.  either sd card has to be initialized earlier or we need a conditional compile for those using sd cards

**Reply by arendst** [reply]
I'll see how to earlianize sd support.

**Reply by arendst** [reply]
With https://github.com/arendst/Tasmota/commit/3edb5bb0964e92a9000cafd8be950974b94067bd I moved up SD card init in the chain.

@gemu2015 can you pls test if this solves the issue?

**Reply by gemu2015** [reply]
OK, works, thx
